release/util.bzl

```diff
@@ -79,3 +82,88 @@ def sdk_tarball(name, version):
         """.format(version = version),
         visibility = ["//visibility:public"],
     )
+
+def _protos_zip_impl(ctx):
+    posix = ctx.toolchains["@rules_sh//sh/posix:toolchain_type"]
+    tmp_dir = ctx.actions.declare_directory("tmp_dir")
+    zipper_args_file = ctx.actions.declare_file(
+        ctx.label.name + ".zipper_args",
+    )
+    tools = [ctx.executable._tar, ctx.executable._gzip]
+    ctx.actions.run_shell(
+        inputs = [ctx.file._ledger_api_tarball] + ctx.files._daml_lf_tarballs,
+        outputs = [tmp_dir],
+        tools = tools,
+        command = """
+          set -eou pipefail
+          export PATH=$PATH:{path}
+          tar xf {ledger_api_tarball} -C {tmp_dir}
+          for file in {lf_tarballs}
+          do
+              tar xf $file -C {tmp_dir}
+          done
+        """.format(
+            ledger_api_tarball = ctx.file._ledger_api_tarball.path,
+            tmp_dir = tmp_dir.path,
+            lf_tarballs = " ".join([f.path for f in ctx.files._daml_lf_tarballs]),
+            path = ":".join(["$PWD/`dirname {tool}`".format(tool = tool.path) for tool in tools]),
+        ),
+    )
+
+    # zipper does not have an option to recursively zip files so we
+    # use find to list the files.
+    ctx.actions.run_shell(
+        outputs = [zipper_args_file],
+        inputs = [tmp_dir],
+        command = """
+        {find} -L {tmp_dir} -type f | {sed} -E 's#^{tmp_dir}/(.*)$#protos-{version}/\\1={tmp_dir}/\\1#' > {args_file}
+        """.format(
+            version = sdk_version,
+            find = posix.commands["find"],
+            sed = posix.commands["sed"],
+            tmp_dir = tmp_dir.path,
+            args_file = zipper_args_file.path,
+        ),
+    )
+    ctx.actions.run(
+        outputs = [ctx.outputs.out],
+        inputs = [zipper_args_file, tmp_dir],
+        executable = ctx.executable._zipper,
+        arguments = ["cC", ctx.outputs.out.path, "@" + zipper_args_file.path],
+    )
+
+protos_zip = rule(
+    implementation = _protos_zip_impl,
+    attrs = {
+        "_daml_lf_tarballs": attr.label_list(
+            allow_files = True,
+            default = [
+                Label("//daml-lf/archive:daml_lf_{}_archive_proto_tarball.tar.gz".format(version))
+                for version in LF_VERSIONS
+            ],
+        ),
+        "_ledger_api_tarball": attr.label(allow_single_file = True, default = Label("//ledger-api/grpc-definitions:ledger-api-protos.tar.gz")),
```

Comment:
These look more like regular attributes than hidden attributes. But, given that this is only used once, it's probably fine.

Reply:
I’ve changed it to regular attributes.

Reply:
Thanks!